### PR TITLE
[W4.3d][W11-A2]Kevin Leonardo

### PR DESCRIPTION
**Pull request by k-l-a**
Pull request for LO W4.3d implement associations.
Split Address class into 4 different classes : Block, Street, Unit, PostalCode.
Classes are implemented simply to hold their respective values, no additional functionality was added.

### CODE REVIEW

File: src/seedu/addressbook/data/person/Address.java [hunk 1]
```diff
@@ -2,17 +2,24 @@
...
  * Represents a Person's address in the address book.
  * Guarantees: immutable; is valid as declared in {@link #isValidAddress(String)}
  */
 public class Address {
 
-    public static final String EXAMPLE = "123, some street";
-    public static final String MESSAGE_ADDRESS_CONSTRAINTS = "Person addresses can be in any format";
-    public static final String ADDRESS_VALIDATION_REGEX = ".+";
+    public static final String EXAMPLE = "123, some street, some unit, some postal code";
+    public static final String MESSAGE_ADDRESS_CONSTRAINTS = "Person addresses should be 4 strings separated by a ', '.";
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Address.java]: good to see you have updated the help message.

File: src/seedu/addressbook/data/person/Address.java [hunk 2]
```diff
@@ -2,17 +2,24 @@
...
+    public static final String ADDRESS_VALIDATION_REGEX = "(.+, ){3}.+";
 
     public final String value;
+    public final Block BLOCK;
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Address.java]: why retaining the `value` here? isn't it sufficient to manage the data as `Block`, `Street`, `Unit`, `Postal Code`? 
Think what can go wrong maintaining duplicate data?

File: src/seedu/addressbook/data/person/Address.java [hunk 3]
```diff
@@ -26,6 +33,11 @@ public Address(String address, boolean isPrivate) throws IllegalValueException {
         if (!isValidAddress(trimmedAddress)) {
             throw new IllegalValueException(MESSAGE_ADDRESS_CONSTRAINTS);
         }
+        List<String> splitAddress = Arrays.asList(trimmedAddress.split(","));
+        this.BLOCK = new Block(splitAddress.get(0));
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Address.java]: What does our coding standard say about using numbers in array indices? [Hint! read magic numbers]

File: src/seedu/addressbook/data/person/Block.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package seedu.addressbook.data.person;
+
+public class Block {
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Block.java]: need to add header comments to the class and its methods

File: src/seedu/addressbook/data/person/Block.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+    
+    public final String value;
+    public Block(String block) {
+        
+        this.value = block;
+    }
```
Comment (okkhoy) [inline src/seedu/addressbook/data/person/Block.java]: can consider moving the validation logic inside here..

File: test/input.txt [hunk 6]
```diff
@@ -28,31 +28,32 @@
...
   
   # should catch invalid person data
-  add []\[;] p/12345 e/valid@e.mail a/valid, address
-  add Valid Name p/not_numbers e/valid@e.mail a/valid, address
-  add Valid Name p/12345 e/notAnEmail a/valid, address
-  add Valid Name p/12345 e/valid@e.mail a/valid, address t/invalid_-[.tag
+  add []\[;] p/12345 e/valid@e.mail a/valid, address, need, four
+  add Valid Name p/not_numbers e/valid@e.mail a/valid, address, need, four
+  add Valid Name p/12345 e/notAnEmail a/valid, address, need, four
+  add Valid Name p/12345 e/valid@e.mail a/valid, address, need, four t/invalid_-[.tag
```
Comment (okkhoy) [inline test/input.txt]: good to see tests updated!

**Comment by okkhoy** [comment]
@k-l-a 
Some comments added. Please ack & close the PR after reading comments.

File: src/seedu/addressbook/data/person/Address.java (hunk 3 above)
Comment (k-l-a) [reply src/seedu/addressbook/data/person/Address.java]: Yes, the number should have been represented by a variable along the lines of BLOCK_POSITION, same for all the other classes (Unit, Street, PostalCode). My bad, missed that.

File: src/seedu/addressbook/data/person/Block.java (hunk 5 above)
Comment (k-l-a) [reply src/seedu/addressbook/data/person/Block.java]: I wanted to put a validation for each of the classes, but didn't have the time.
The validation change in Address is just checking if the input can be split into 4.

**Comment by k-l-a** [comment]
Thank you for the review, PR will be closed now.

**Pull request closed without merging**